block interval 200ms - spurious view changes

Seen on securekg conodes, I tried to make a block interval of 200ms, and it seemed to work, but then died with:

```
2 : (    byzcoin.(*Service).sendViewChangeReq: 120) - tls://localhost:18004 sending view-change request for view: ID: e5067520fb4f0ae9a8dc834adeb629f70d2f9b990316a541ddae611838394670, Gen: d41a65bd68911e848e9c40ad7053a66b01946e66e96218a8ebb3b7ccf4b81ae1, Leader: 1
panic: heartbeat monitors are started after the creation of the genesis block, so the block should always exist

goroutine 35 [running]:
github.com/dedis/cothority/byzcoin.(*Service).monitorLeaderFailure.func1(0xc4200d4c60)
	/go/src/github.com/dedis/cothority/byzcoin/service.go:1844 +0x43a
created by github.com/dedis/cothority/byzcoin.(*Service).monitorLeaderFailure
	/go/src/github.com/dedis/cothority/byzcoin/service.go:1836 +0xc5
W : ( viewchange.(*Controller).processAnomaly: 253) - Controller is not accepting anomalies for earlier views
```

--- a/byzcoin/service.go
+++ b/byzcoin/service.go
@@ -2183,22 +2183,34 @@ func (s *Service) monitorLeaderFailure() {
 			case key := <-s.heartbeatsTimeout:
 				log.Lvlf3("%s: missed heartbeat for %x", s.ServerIdentity(), key)
 				gen := []byte(key)
-				latest, err := s.db().GetLatestByID(gen)
-				if err != nil {
-					log.Error("heartbeat monitors are started after "+
-						"the creation of the genesis block, "+
-						"so the block should always exist: ", err)
+
+				genBlock := s.db().GetByID(gen)
+				if genBlock == nil {
+					// This should not happen as the heartbeats are started after
+					// a new skipchain is created or when the conode starts ..
+					log.Error("heartbeat monitors are started after " +
+						"the creation of the genesis block, " +
+						"so the block should always exist")
+					// .. but just in case we stop the heartbeat
 					s.heartbeats.stop(key)
 				}
-				req := viewchange.InitReq{
-					SignerID: s.ServerIdentity().ID,
-					View: viewchange.View{
-						ID:          latest.Hash,
-						Gen:         gen,
-						LeaderIndex: 1,
-					},
+
+				latest, err := s.db().GetLatestByID(gen)
+				if err != nil {
+					log.Errorf("failed to get the latest block: %v", err)
+				} else {
+					// Send only if the latest block is consistent as it wouldn't
+					// anyway if we're out of sync with the chain
+					req := viewchange.InitReq{
+						SignerID: s.ServerIdentity().ID,
+						View: viewchange.View{
+							ID:          latest.Hash,
+							Gen:         gen,
+							LeaderIndex: 1,
+						},
+					}
+					s.viewChangeMan.addReq(req)
 				}
-				s.viewChangeMan.addReq(req)
 			case <-s.closeLeaderMonitorChan:
 				log.Lvl2(s.ServerIdentity(), "closing heartbeat timeout monitor")
 				return

--- a/byzcoin/viewchange_test.go
+++ b/byzcoin/viewchange_test.go
@@ -232,3 +232,20 @@ func TestViewChange_LostSync(t *testing.T) {
 		require.True(t, leader.Equal(s.services[3].ServerIdentity()))
 	}
 }
+
+func TestViewChange_MonitorFailure(t *testing.T) {
+	s := newSerN(t, 1, time.Second, 3, true)
+	defer s.local.CloseAll()
+
+	log.OutputToBuf()
+	defer log.OutputToOs()
+
+	// heartbeats an unknown skipchain: this should NOT panic or crash
+	s.service().heartbeatsTimeout <- "abc"
+
+	time.Sleep(1 * time.Second)
+
+	stderr := log.GetStdErr()
+	require.Contains(t, stderr, "heartbeat monitors are started after the creation")
+	require.Contains(t, stderr, "failed to get the latest block")
+}

